src/ffi.rs

```diff
@@ -40,3 +44,248 @@ pub fn disable_internal_nscd() {
+/// `libc::hostent` is set to null by glibc.
+fn unmarshal_gethostbyxx(
+    hostent: *mut libc::hostent,
+    herrno: libc::c_int,
+) -> anyhow::Result<Hostent> {
+    if !hostent.is_null() {
+        let res = from_libc_hostent(unsafe { *hostent } )?;
+        Ok(res)
+    } else {
+        Ok(
+            // This is a default hostent header we're supposed to use
+            // to convey a lookup error. This is a glibc quirk, I have
+            // nothing to do with that, don't blame me :)
+            Hostent {
+                name: "".to_string(),
+                aliases: Vec::new(),
+                addr_type: -1,
+                addr_list: Vec::new(),
+                herrno,
+            },
+        )
+    }
+}
+
+pub fn gethostbyaddr_r(addr: LibcIp) -> anyhow::Result<Hostent> {
+    let (addr, len, af) = match addr {
+        LibcIp::V4(ref ipv4) => (ipv4 as &[u8], 4, libc::AF_INET),
+        LibcIp::V6(ref ipv6) => (ipv6 as &[u8], 16, libc::AF_INET6),
+    };
+
+    let mut ret_hostent: libc::hostent = libc::hostent {
+        h_name: ptr::null_mut(),
+        h_aliases: ptr::null_mut(),
+        h_addrtype: 0,
+        h_length: 0,
+        h_addr_list: ptr::null_mut(),
+    };
+    let mut herrno: libc::c_int = 0;
+    let mut hostent_result = ptr::null_mut();
+    // We start with a 1024 bytes buffer, the nscd default. See
+    // scratch_buffer.h in the glibc codebase
+    let mut buf: Vec<u8> = Vec::with_capacity(1024);
+    loop {
+        let ret = unsafe {
+            glibcffi::gethostbyaddr_r(
+                addr.as_ptr() as *const libc::c_void,
+                len,
+                af,
+                &mut ret_hostent,
+                buf.as_mut_ptr() as *mut libc::c_char,
+                (buf.capacity() as size_t).try_into().unwrap(),
+                &mut hostent_result,
+                &mut herrno,
+            )
+        };
+
+        if ret == libc::ERANGE && buf.capacity() < 10 * 1000 * 1000  {
+            buf.reserve(buf.capacity() * 2);
+        } else {
+            break;
+        }
+    }
+    unmarshal_gethostbyxx(hostent_result, herrno)
+}
+
+/// Typesafe wrapper around the gethostbyname2_r glibc function
+///
+/// af is either nix::libc::AF_INET or nix::libc::AF_INET6
+pub fn gethostbyname2_r(name: String, af: libc::c_int) -> anyhow::Result<Hostent> {
+    let name = CString::new(name).unwrap();
+
+    // Prepare a libc::hostent and the pointer to the result list,
+    // which will be passed to the glibcffi::gethostbyname2_r call.
+    let mut ret_hostent: libc::hostent = libc::hostent {
+        h_name: ptr::null_mut(),
+        h_aliases: ptr::null_mut(),
+        h_addrtype: 0,
+        h_length: 0,
+        h_addr_list: ptr::null_mut(),
+    };
+    let mut herrno: libc::c_int = 0;
+    // The 1024 initial size comes from the Glibc default. It fit most
+    // of the requests in practice.
+    let mut buf: Vec<u8> = Vec::with_capacity(1024);
```

Comment:
I think by the book this may need to be `MaybeUninit<u8>` even though it's never read. Let me consult with @geofft.

Reply:
We chatted and I think I'm being overly cautious here.